src/conn/conn_dhandle.c

```diff
@@ -180,45 +181,87 @@ __wt_conn_btree_sync_and_close(WT_SESSION_IMPL *session, bool final, bool force)
 	 */
 	__wt_spin_lock(session, &dhandle->close_lock);
 
-	/*
-	 * The close can fail if an update cannot be written, return the EBUSY
-	 * error to our caller for eventual retry.
-	 *
-	 * If we are forcing the close, just mark the handle dead and the tree
-	 * will be discarded later.  Don't do this for memory-mapped trees: we
-	 * have to close the file handle to allow the file to be removed, but
-	 * memory mapped trees contain pointers into memory that will become
-	 * invalid if the mapping is closed.
-	 */
+	/* Reset the tree's eviction priority (if any). */
+	__wt_evict_priority_clear(session);
+
+	discard = marked_dead = false;
 	if (!F_ISSET(btree,
 	    WT_BTREE_SALVAGE | WT_BTREE_UPGRADE | WT_BTREE_VERIFY)) {
-		if (force && (bm == NULL || !bm->is_mapped(bm, session))) {
-			F_SET(session->dhandle, WT_DHANDLE_DEAD);
+		/*
+		 * If the object is already marked dead, we're just here to
+		 * discard it.
+		 */
+		if (F_ISSET(dhandle, WT_DHANDLE_DEAD))
+			discard = true;
+
+		/*
+		 * If a standard object and it's OK to mark the handle dead
```

Comment:
I find the phrase "If a standard object" awkward.  The meaning of "standard object" isn't clear nor is it obvious in the conditional below what that would mean. (And it probably should say "If this is a...").  The "it's OK to mark..." refers to a 5-part conditional.  I'd prefer something more obvious like "It is okay to make the handle dead only when..."